pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -642,9 +642,86 @@ impl RemoteTimelineClient {
+        new_parent: &TimelineId,
+    ) -> anyhow::Result<()> {
+        // FIXME: because of how Timeline::schedule_uploads works when called from layer flushing
+        // and reads the in-memory part we cannot do the detaching like this
+        let receiver = {
+            let mut guard = self.upload_queue.lock().unwrap();
+            let upload_queue = guard.initialized_mut()?;
+
+            upload_queue.latest_metadata.reparent(new_parent);
+
+            self.schedule_index_upload(upload_queue, upload_queue.latest_metadata.clone());
+
+            self.schedule_barrier0(upload_queue)
+        };
+
+        Self::wait_completion0(receiver).await
+    }
+
+    /// Schedule an index part update as the last step of timeline detach.
+    ///
+    /// Metadata of adoption is added to the next index part recording the path for WAL based
+    /// disaster recovery.
     ///
-    /// Launch an upload operation in the background.
+    /// The returned future waits for the upload to complete, which means that remote storage is
+    /// now in sync, and the timeline can be separated from its ancestor. The next time the
+    /// timeline is loaded from remote storage, it will have no ancestor.
     ///
+    /// Cancellation-safety: cancelling or not waiting for the returned future will not cancel the
+    /// upload.
+    pub(crate) async fn schedule_detaching_from_ancestor_and_wait(
+        self: &Arc<Self>,
+        adopted: (TimelineId, Lsn),
+    ) -> anyhow::Result<()> {
+        let receiver = {
+            let mut guard = self.upload_queue.lock().unwrap();
+            let upload_queue = guard.initialized_mut()?;
+
+            // TODO: update lineage in index_part once we have it
+            upload_queue
+                .latest_metadata
+                .detach_from_ancestor(&adopted.0, &adopted.1);
+
+            self.schedule_index_upload(upload_queue, upload_queue.latest_metadata.clone());
+
+            self.schedule_barrier0(upload_queue)
+        };
+
+        Self::wait_completion0(receiver).await
+    }
+
+    /// Schedules uploading a new version of `index_part.json` with the given layers added and
+    /// waits for it to complete.
+    ///
+    /// This is used with `Timeline::detach_ancestor` functionality.
+    pub(crate) async fn schedule_adding_existing_layers_to_index_upload_and_wait(
+        self: &Arc<Self>,
+        layers: &[Layer],
+    ) -> anyhow::Result<()> {
+        let barrier = {
+            let mut guard = self.upload_queue.lock().unwrap();
+            let upload_queue = guard.initialized_mut()?;
+
+            for layer in layers {
+                upload_queue
+                    .latest_files
+                    .insert(layer.layer_desc().filename(), layer.metadata());
+            }
+
+            let barrier = self.schedule_barrier0(upload_queue);
+            self.launch_queued_tasks(upload_queue);
+            barrier
+        };
+
+        Self::wait_completion0(barrier).await
+    }
```

Comment:
Why is this not re-using `schedule_layer_file_upload0`?

Reply:
Because we don't want to make it visible in "prepare" phase before the "completion" stage:

1. schedule_layer_file_upload0 creates an op, updates metrics, pushes the op to queue
2. it is per layer
3. this is one per all of the prepared layers